presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java

```diff
@@ -0,0 +1,250 @@
+            name = DereferenceExpression.getQualifiedName((DereferenceExpression) expression);
+        }
+        return name;
+    }
+
+    public Optional<ResolvedField> tryResolveField(Expression node, QualifiedName name)
+    {
+        return resolveField(node, name, true);
+    }
+
+    private Optional<ResolvedField> resolveField(Expression node, QualifiedName name, boolean local)
+    {
+        List<Field> matches = relation.resolveFields(name);
+        if (matches.size() > 1) {
+            throw createAmbiguousAttributeException(node, name);
+        }
+
+        if (matches.isEmpty()) {
+            if (isColumnReference(name, relation)) {
+                return Optional.empty();
+            }
+            Scope boundary = this;
+            while (!boundary.queryBoundary) {
+                if (boundary.parent.isPresent()) {
+                    boundary = boundary.parent.get();
+                }
+                else {
+                    return Optional.empty();
+                }
+            }
+            if (boundary.parent.isPresent()) {
+                // jump over the query boundary
+                return boundary.parent.get().resolveField(node, name, false);
+            }
+            return Optional.empty();
+        }
+        else {
+            return Optional.of(asResolvedField(getOnlyElement(matches), local));
+        }
+    }
+
+    private ResolvedField asResolvedField(Field field, boolean local)
+    {
+        int fieldIndex = relation.indexOf(field);
+        return new ResolvedField(this, field, fieldIndex, local);
+    }
+
+    public boolean isColumnReference(QualifiedName name)
+    {
+        Scope current = this;
+        while (current != null) {
+            if (isColumnReference(name, current.relation)) {
+                return true;
+            }
+
+            current = current.parent.orElse(null);
+        }
+
+        return false;
+    }
+
+    private boolean isColumnReference(QualifiedName name, RelationType relation)
+    {
+        while (name.getPrefix().isPresent()) {
+            name = name.getPrefix().get();
+            if (!relation.resolveFields(name).isEmpty()) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public Optional<WithQuery> getNamedQuery(String name)
+    {
+        if (namedQueries.containsKey(name)) {
+            return Optional.of(namedQueries.get(name));
+        }
+
+        if (parent.isPresent()) {
+            return parent.get().getNamedQuery(name);
+        }
+
+        return Optional.empty();
+    }
+
+    public boolean isApproximate()
+    {
+        return approximate;
+    }
+
+    public Optional<Scope> getLateralScope()
+    {
+        return lateralScope;
+    }
+
+    public static final class Builder
+    {
+        private RelationType relationType = new RelationType();
+        private Optional<Boolean> approximate = Optional.empty();
```

Comment:
Optional<Boolean> looks suspicious. In this case we don't use the third (empty) value of this tri-state logic, we just fall back to false upon build. Why not handle this using a field initializer?

`boolean approximate = false`